lib/auth.js

```diff
@@ -0,0 +1,54 @@
+import React, { useState, useEffect, useContext, createContext } from 'react';
+import firebase from './firebase';
+
+const authContext = createContext();
+
+export function ProvideAuth({ children }) {
+  const auth = useProvideAuth();
+  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
+}
+
+export const useAuth = () => {
+  return useContext(authContext);
+};
+
+function useProvideAuth() {
+  const [user, setUser] = useState(null);
+
+  const signinWithGithub = () => {
+    return firebase
+      .auth()
+      .signInWithPopup(new firebase.auth.GithubAuthProvider())
+      .then((response) => {
+        setUser(response.user);
+        return response.user;
+      });
+  };
+
+  const signout = () => {
+    return firebase
+      .auth()
+      .signOut()
+      .then(() => {
+        setUser(false);
+      });
+  };
+
+  useEffect(() => {
+    const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
+      if (user) {
+        setUser(user);
+      } else {
+        setUser(false);
```

Comment:
@leerob was curious why `setUser(false)` and not `setUser(null)`?

Reply:
Good question! Either could work 😄